pkg/util/gziphandler/gzip.go

```diff
@@ -118,19 +116,21 @@ func (w *GzipResponseWriter) Write(b []byte) (int, error) {
 		ce    = w.Header().Get(contentEncoding)
 	)
 	// Only continue if they didn't already choose an encoding or a known unhandled content length or type.
-	if ce == "" && (cl == 0 || cl >= w.minSize) && (ct == "" || handleContentType(w.contentTypes, ct)) {
+	if ce == "" && (cl == 0 || cl >= w.minSize || !w.acceptsIdentity) && (ct == "" || handleContentType(w.contentTypes, ct)) {
```

Comment:
I know you just appended here, but the double negatives in these comments makes intent hard to follow

Reply:
I completely agree, but I wanted to keep the code changes at minimum compared to the original code. Maybe I should fix that code.

Reply:
I finally refactored this, trying to make it cleaner. I did it in multiple commits to make the reviewing (hopefully) easier, PTAL:
- https://github.com/grafana/mimir/pull/1864/commits/d27a4eb59956ce08e5db09349e3532621ef3b461
- https://github.com/grafana/mimir/pull/1864/commits/a8673b307c87f6aa67280a61682fc8355f26243a
- https://github.com/grafana/mimir/pull/1864/commits/d22899c1c8d01fda74278a48db653bd050dcf669
- https://github.com/grafana/mimir/pull/1864/commits/705540dcb40c16de631b778770a53ccc6b079291
- https://github.com/grafana/mimir/pull/1864/commits/a21356223b79d9041c66b9049540f09768adab27

Reply:
@johannaratliff I'll merge this as it is now, but I'm open to any extra feedback you may have.